File: app/src/features/workspaces/model/permissions.ts
import type {
  PermissionOverridePayload,
  WorkspaceRoleResponse,
} from '@/entities/workspace/api'

export type BaseRole = 'viewer' | 'editor' | 'admin'
export type SystemRole = BaseRole | 'owner'
export type InviteRoleValue = `system:${SystemRole}` | `custom:${string}`

export const PERMISSION_GROUPS = [
  {
    title: 'Documents & Folders',
    permissions: [
      { id: 'doc:view', label: 'View documents' },
      { id: 'doc:edit', label: 'Edit documents' },
      { id: 'doc:create', label: 'Create documents' },
      { id: 'doc:archive', label: 'Archive documents' },
      { id: 'doc:delete', label: 'Delete documents' },
      { id: 'doc:move', label: 'Move documents' },
      { id: 'folder:create', label: 'Create folders' },
      { id: 'folder:delete', label: 'Delete folders' },
    ],
  },
  {
    title: 'Files & Attachments',
    permissions: [
      { id: 'file:upload', label: 'Upload files' },
      { id: 'file:delete', label: 'Delete files' },
    ],
  },
  {
    title: 'Sharing & Publishing',
    permissions: [
      { id: 'share:create', label: 'Create share links' },
      { id: 'share:delete', label: 'Delete share links' },
      { id: 'public:publish', label: 'Publish publicly' },
      { id: 'public:unpublish', label: 'Unpublish' },
    ],
  },
  {
    title: 'Members & Workspaces',
    permissions: [
      { id: 'member:view', label: 'View members' },
      { id: 'member:invite', label: 'Invite members' },
      { id: 'member:update_role', label: 'Manage roles' },
      { id: 'member:remove', label: 'Remove members' },
      { id: 'workspace:update_settings', label: 'Update workspace settings' },
      { id: 'workspace:delete', label: 'Delete workspace' },
    ],
  },
  {
    title: 'Integrations',
    permissions: [
      { id: 'plugin:install', label: 'Install plugins' },
      { id: 'plugin:uninstall', label: 'Uninstall plugins' },
      { id: 'plugin:run', label: 'Run plugins' },
      { id: 'git:init', label: 'Initialize Git' },
      { id: 'git:sync', label: 'Sync with Git' },
      { id: 'git:configure', label: 'Configure Git' },
    ],
  },
  {
    title: 'Automation & API',
    permissions: [
      { id: 'shortcut:update', label: 'Manage shortcuts' },
      { id: 'api_token:manage', label: 'Manage API tokens' },
    ],
  },
] as const

export const ALL_PERMISSIONS = PERMISSION_GROUPS.flatMap((group) =>
  group.permissions.map((item) => item.id),
)

const SYSTEM_ROLE_DEFAULTS: Record<SystemRole, string[]> = {
  owner: ALL_PERMISSIONS,
  admin: ALL_PERMISSIONS.filter((perm) => perm !== 'workspace:delete'),
  editor: [
    'doc:view',
    'doc:edit',
    'doc:create',
    'doc:archive',
    'doc:delete',
    'doc:move',
    'folder:create',
    'folder:delete',
    'file:upload',
    'file:delete',
    'share:create',
    'share:delete',
    'public:publish',
    'public:unpublish',
    'plugin:run',
    'plugin:install',
    'plugin:uninstall',
    'git:init',
    'git:sync',
    'git:configure',
    'shortcut:update',
    'api_token:manage',
  ],
  viewer: ['doc:view'],
}

export const DEFAULT_INVITE_ROLE: InviteRoleValue = 'system:editor'

export function buildSystemPermissionSet(role: SystemRole) {
  return new Set(SYSTEM_ROLE_DEFAULTS[role])
}

export function applyOverrides(
  base: Set<string>,
  overrides?: PermissionOverridePayload[] | null,
) {
  if (!overrides || overrides.length === 0) {
    return base
  }
  const next = new Set(base)
  overrides.forEach((override) => {
    if (override.allowed) {
      next.add(override.permission)
    } else {
      next.delete(override.permission)
    }
  })
  return next
}

export function roleToPermissionSet(role: WorkspaceRoleResponse) {
  const base = buildSystemPermissionSet(role.base_role as BaseRole)
  return applyOverrides(base, role.overrides)
}

export function computeOverrides(baseRole: BaseRole, selected: Set<string>) {
  const baseSet = buildSystemPermissionSet(baseRole)
  const overrides: PermissionOverridePayload[] = []
  ALL_PERMISSIONS.forEach((perm) => {
    const inBase = baseSet.has(perm)
    const isSelected = selected.has(perm)
    if (inBase !== isSelected) {
      overrides.push({ permission: perm, allowed: isSelected })
    }
  })
  return overrides
}

File: api/crates/domain/src/workspaces/permissions.rs
use std::collections::BTreeSet;

pub const PERM_DOC_VIEW: &str = "doc:view";
pub const PERM_DOC_EDIT: &str = "doc:edit";
pub const PERM_DOC_CREATE: &str = "doc:create";
pub const PERM_DOC_ARCHIVE: &str = "doc:archive";
pub const PERM_DOC_DELETE: &str = "doc:delete";
pub const PERM_DOC_MOVE: &str = "doc:move";
pub const PERM_FOLDER_CREATE: &str = "folder:create";
pub const PERM_FOLDER_DELETE: &str = "folder:delete";
pub const PERM_FILE_UPLOAD: &str = "file:upload";
pub const PERM_FILE_DELETE: &str = "file:delete";
pub const PERM_SHARE_CREATE: &str = "share:create";
pub const PERM_SHARE_DELETE: &str = "share:delete";
pub const PERM_PUBLIC_PUBLISH: &str = "public:publish";
pub const PERM_PUBLIC_UNPUBLISH: &str = "public:unpublish";
pub const PERM_MEMBER_VIEW: &str = "member:view";
pub const PERM_MEMBER_INVITE: &str = "member:invite";
pub const PERM_MEMBER_UPDATE_ROLE: &str = "member:update_role";
pub const PERM_MEMBER_REMOVE: &str = "member:remove";
pub const PERM_WORKSPACE_UPDATE: &str = "workspace:update_settings";
pub const PERM_WORKSPACE_DELETE: &str = "workspace:delete";
pub const PERM_PLUGIN_INSTALL: &str = "plugin:install";
pub const PERM_PLUGIN_UNINSTALL: &str = "plugin:uninstall";
pub const PERM_PLUGIN_RUN: &str = "plugin:run";
pub const PERM_GIT_INIT: &str = "git:init";
pub const PERM_GIT_SYNC: &str = "git:sync";
pub const PERM_GIT_CONFIGURE: &str = "git:configure";
pub const PERM_SHORTCUT_UPDATE: &str = "shortcut:update";
pub const PERM_API_TOKEN_MANAGE: &str = "api_token:manage";

const ALL_PERMISSIONS: &[&str] = &[
    PERM_DOC_VIEW,
    PERM_DOC_EDIT,
    PERM_DOC_CREATE,
    PERM_DOC_ARCHIVE,
    PERM_DOC_DELETE,
    PERM_DOC_MOVE,
    PERM_FOLDER_CREATE,
    PERM_FOLDER_DELETE,
    PERM_FILE_UPLOAD,
    PERM_FILE_DELETE,
    PERM_SHARE_CREATE,
    PERM_SHARE_DELETE,
    PERM_PUBLIC_PUBLISH,
    PERM_PUBLIC_UNPUBLISH,
    PERM_MEMBER_VIEW,
    PERM_MEMBER_INVITE,
    PERM_MEMBER_UPDATE_ROLE,
    PERM_MEMBER_REMOVE,
    PERM_WORKSPACE_UPDATE,
    PERM_WORKSPACE_DELETE,
    PERM_PLUGIN_INSTALL,
    PERM_PLUGIN_UNINSTALL,
    PERM_PLUGIN_RUN,
    PERM_GIT_INIT,
    PERM_GIT_SYNC,
    PERM_GIT_CONFIGURE,
    PERM_SHORTCUT_UPDATE,
    PERM_API_TOKEN_MANAGE,
];

#[derive(Debug, Clone, Default)]
pub struct PermissionSet {
    allowed: BTreeSet<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PermissionOverride {
    pub permission: String,
    pub allowed: bool,
}

impl PermissionOverride {
    pub fn new(permission: impl Into<String>, allowed: bool) -> Self {
        Self {
            permission: permission.into(),
            allowed,
        }
    }
}

impl PermissionSet {
    pub fn is_empty(&self) -> bool {
        self.allowed.is_empty()
    }

    pub fn allows(&self, permission: &str) -> bool {
        self.allowed.contains(permission)
    }

    pub fn insert(&mut self, permission: impl Into<String>) {
        self.allowed.insert(permission.into());
    }

    pub fn remove(&mut self, permission: &str) {
        self.allowed.remove(permission);
    }

    pub fn iter(&self) -> impl Iterator<Item = &String> {
        self.allowed.iter()
    }

    pub fn to_vec(&self) -> Vec<String> {
        self.allowed.iter().cloned().collect()
    }

    pub fn contains_all(&self, other: &PermissionSet) -> bool {
        other
            .allowed
            .iter()
            .all(|permission| self.allowed.contains(permission))
    }

    pub fn union(mut self, other: PermissionSet) -> Self {
        for perm in other.allowed {
            self.allowed.insert(perm);
        }
        self
    }

    pub fn from_slice(perms: &[&str]) -> Self {
        let mut set = PermissionSet::default();
        for perm in perms {
            set.allowed.insert(perm.to_string());
        }
        set
    }

    pub fn from_strings<I, S>(perms: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        let mut set = PermissionSet::default();
        for perm in perms {
            set.allowed.insert(perm.into());
        }
        set
    }

    pub fn all() -> Self {
        Self::from_slice(ALL_PERMISSIONS)
    }
}

pub fn system_role_permissions(role: &str) -> PermissionSet {
    match role {
        "owner" => PermissionSet::all(),
        "admin" => {
            let mut base = PermissionSet::all();
            base.remove(PERM_WORKSPACE_DELETE);
            base
        }
        "editor" => PermissionSet::from_slice(&[
            PERM_DOC_VIEW,
            PERM_DOC_EDIT,
            PERM_DOC_CREATE,
            PERM_DOC_ARCHIVE,
            PERM_DOC_DELETE,
            PERM_DOC_MOVE,
            PERM_FOLDER_CREATE,
            PERM_FOLDER_DELETE,
            PERM_FILE_UPLOAD,
            PERM_FILE_DELETE,
            PERM_SHARE_CREATE,
            PERM_SHARE_DELETE,
            PERM_PUBLIC_PUBLISH,
            PERM_PUBLIC_UNPUBLISH,
            PERM_PLUGIN_RUN,
            PERM_PLUGIN_INSTALL,
            PERM_PLUGIN_UNINSTALL,
            PERM_GIT_INIT,
            PERM_GIT_SYNC,
            PERM_GIT_CONFIGURE,
            PERM_SHORTCUT_UPDATE,
            PERM_API_TOKEN_MANAGE,
        ]),
        "viewer" => PermissionSet::from_slice(&[PERM_DOC_VIEW]),
        _ => PermissionSet::default(),
    }
}

pub fn apply_custom_overrides<I>(mut base: PermissionSet, overrides: I) -> PermissionSet
where
    I: IntoIterator<Item = PermissionOverride>,
{
    for item in overrides {
        if item.allowed {
            base.insert(item.permission);
        } else {
            base.remove(&item.permission);
        }
    }
    base
}
